fedimint-recurringd/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+use fedimint_mint_client::MintClientInit;
+use futures::StreamExt;
+use lightning_invoice::{Bolt11Invoice, Bolt11InvoiceDescription, Sha256};
+use lnurl::lnurl::LnUrl;
+use lnurl::pay::{LnURLPayInvoice, PayResponse};
+use lnurl::Tag;
+use tokio::sync::{Notify, RwLock};
+use tracing::{info, warn};
+
+use crate::db::{
+    load_federation_client_databases, open_client_db, try_add_federation_database,
+    FederationDbPrefix, PaymentCodeEntry, PaymentCodeInvoiceEntry, PaymentCodeInvoiceKey,
+    PaymentCodeKey, PaymentCodeNextInvoiceIndexKey, PaymentCodeVariant,
+};
+
+mod db;
+
+#[derive(Clone)]
+pub struct RecurringInvoiceServer {
+    db: Database,
+    clients: Arc<RwLock<HashMap<FederationId, ClientHandleArc>>>,
+    invoice_generated: Arc<Notify>,
+    base_url: SafeUrl,
+}
+
+impl RecurringInvoiceServer {
+    pub async fn new(db: impl IRawDatabase + 'static, base_url: SafeUrl) -> anyhow::Result<Self> {
+        let db = Database::new(db, Default::default());
+
+        let mut clients = HashMap::<_, ClientHandleArc>::new();
+
+        for (federation_id, db) in load_federation_client_databases(&db).await {
+            let mut client_builder = Client::builder(db).await?;
+            client_builder.with_module(LightningClientInit::default());
+            client_builder.with_module(MintClientInit);
+            client_builder.with_primary_module_kind(ModuleKind::from_static_str("mint"));
+            let client = client_builder.open(Self::default_secret()).await?;
+            clients.insert(federation_id, Arc::new(client));
+        }
+
+        Ok(Self {
+            db,
+            clients: Arc::new(RwLock::new(clients)),
+            invoice_generated: Arc::new(Default::default()),
+            base_url,
+        })
+    }
+
+    /// We don't want to hold any money or sign anything ourselves, we only use
+    /// the client with externally supplied key material and to track
+    /// ongoing progress of other users' receives.
+    fn default_secret() -> DerivableSecret {
+        DerivableSecret::new_root(&[], &[])
+    }
+
+    pub async fn register_federation(
+        &self,
+        invite_code: &InviteCode,
+    ) -> Result<FederationId, RecurringPaymentError> {
+        let federation_id = invite_code.federation_id();
+        info!("Registering federation {}", federation_id);
+
+        // We lock to prevent parallel join attempts
+        // TODO: lock per federation
+        let mut clients = self.clients.write().await;
+        if clients.contains_key(&federation_id) {
+            return Err(RecurringPaymentError::FederationAlreadyRegistered(
+                federation_id,
+            ));
+        }
+
+        // We don't know if joining will succeed or be interrupted. We use a random DB
+        // prefix to initialize the client and only write the prefix to the DB if that
+        // succeeds. If it fails we end up with some orphaned data in the DB, if it ever
+        // becomes a problem we can clean it up later.
```

Comment:
Could you elaborate on this failure mode? How do we expect the `join_federation_static` to fail, but the database end up with orphaned data? Shouldn't the database transaction also fail if the join failed?

I think we should just use the `FederationId` here instead of a random identifier.

Reply:
The client builder doesn't accept a DB transaction, but a whole DB as an argument and when calling `join` does something unspecified with it. The API contract doesn't give me any guarantees about the state of the DB after an interrupted `join` call and e.g. if there were multiple DBTX the first could succeed while the second gets cancelled because of a crash.